### PR TITLE
Multiple list nodes

### PR DESCRIPTION
**Pull request by liranoz12**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/45/ for details (Mellanox internal link).

File: src/vma/iomux/epoll_wait_call.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ int epoll_wait_call::get_current_events()
 		return m_n_all_ready_fds;
 	}
 
-	vector<socket_fd_api *> socket_fd_vec;
+	vma_list_t<socket_fd_api, socket_fd_api::socket_fd_vec_node_offset> socket_fd_vec;
```
Comment (OphirMunk) [inline src/vma/iomux/epoll_wait_call.cpp]: For all cases of ..._socket_fd_ver_... names - change to ..._socket_fs_list_... names.
For example: change variable name socket_fd_vec to socket_fd_list

File: src/vma/iomux/epfd_info.cpp [hunk 2]
```diff
@@ -510,7 +519,9 @@ int epfd_info::mod_fd(int fd, epoll_event *event)
 	}
 
 	if(event->events == 0 || events == 0){
-		m_ready_fds.erase(fd);
+		socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
```
Comment (OphirMunk) [inline src/vma/iomux/epfd_info.cpp]: Missing check: sock_fd->ep_ready_fd_node.is_list_member() ...

File: src/vma/iomux/epfd_info.cpp [hunk 3]
```diff
@@ -595,25 +606,26 @@ void epfd_info::insert_epoll_event_cb(int fd, uint32_t event_flags)
...
-	ep_ready_fd_map_t::iterator iter = m_ready_fds.find(fd);
-	if (iter != m_ready_fds.end()) {
-		iter->second &= ~event_flags;
-	}
-	if (iter->second == 0) {
-		m_ready_fds.erase(iter);
+	socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
+	if (sock_fd->ep_ready_fd_node.is_list_member()) {
+		sock_fd->m_epoll_event_flags &= ~event_flags;
+		if (sock_fd->m_epoll_event_flags == 0) {
```
Comment (OphirMunk) [inline src/vma/iomux/epfd_info.cpp]: In original implementation:
if (iter->second == 0) {
		m_ready_fds.erase(iter);
is outside the if statement:
if (iter != m_ready_fds.end()) {...
while in your new implementation:
if (sock_fd->m_epoll_event_flags == 0) {
is inside the if statement
if (sock_fd->ep_ready_fd_node.is_list_member()) {

Please justify this change in logic

File: src/vma/iomux/epoll_wait_call.cpp [hunk 4]
```diff
@@ -85,45 +85,45 @@ int epoll_wait_call::get_current_events()
...
 	socket_fd_api *p_socket_object;
 	epoll_fd_rec fd_rec;
-	ep_ready_fd_map_t::iterator iter = m_epfd_info->m_ready_fds.begin();
+	list_iterator_t<socket_fd_api, socket_fd_api::ep_ready_fd_node_offset> iter = m_epfd_info->m_ready_fds.begin();
 	while (iter != m_epfd_info->m_ready_fds.end() && i < m_maxevents) {
-		ep_ready_fd_map_t::iterator iter_cpy = iter; // for protection needs
+		list_iterator_t<socket_fd_api, socket_fd_api::ep_ready_fd_node_offset> iter_cpy = iter; // for protection needs
 		++iter;
-		p_socket_object = fd_collection_get_sockfd(iter_cpy->first);
+		p_socket_object = iter_cpy.getPtr();
```
Comment (OphirMunk) [inline src/vma/iomux/epoll_wait_call.cpp]: In original code you get p_socket_object by: 
fd_collection_get_sockfd(iter_cpy->first);
In your code you get p_socket_object by:
iter_cpy.getPtr();

Can you think of a way of getting fd from iter_cpy, then align with the original code and get p_socket_object by: 
fd_collection_get_sockfd(fd);....

File: src/vma/iomux/epfd_info.cpp (hunk 3 above)
Comment (liranoz12) [inline src/vma/iomux/epfd_info.cpp]: The purpose of the second if is to remove the fd from m_ready_fds if the event_flags value is 0.
Like your comment above, we have to check if the fd is a list member before removing it from the list.

> 
socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
if (sock_fd->ep_ready_fd_node.is_list_member()) {
      sock_fd->m_epoll_event_flags &= ~event_flags;
}
if (sock_fd->ep_ready_fd_node.is_list_member() && sock_fd->m_epoll_event_flags == 0) {
m_ready_fds.erase(sock_fd);
}

This is equal to my implementation - insert the second if into the first if and remove the is_list_member() check from the if statement:

> 
socket_fd_api* sock_fd = fd_collection_get_sockfd(fd);
if (sock_fd->ep_ready_fd_node.is_list_member()) {
sock_fd->m_epoll_event_flags &= ~event_flags;
if (sock_fd->m_epoll_event_flags == 0) {
m_ready_fds.erase(sock_fd);
}
}

File: src/vma/iomux/epoll_wait_call.cpp (hunk 1 above)
Comment (liranoz12) [inline src/vma/iomux/epoll_wait_call.cpp]: I will fix it in the next pull request.

File: src/vma/iomux/epfd_info.cpp (hunk 2 above)
Comment (liranoz12) [inline src/vma/iomux/epfd_info.cpp]: I will fix it in the next pull request.

**Pull request closed without merging**